Web Inspector: Wrong breakpoint and search result positions after pretty-printing inline script with multi-line template literal

rcaliman-apple opened the pull request:
#### b21fb0d39d87fb3a040f2b9e962fb7c43e575fe2
<pre>
Web Inspector: Wrong breakpoint and search result positions after pretty-printing inline script with multi-line template literal
<a href="https://bugs.webkit.org/show_bug.cgi?id=291140">https://bugs.webkit.org/show_bug.cgi?id=291140</a>
<a href="https://rdar.apple.com/29417859">rdar://29417859</a>

Reviewed by Devin Rousso.

`JSFormatter.prototype._handleTokenAtNode()` appended multi-line template literals
as a single token to the formatted content. This assumes the token is on a single line.

This actually triggered the assertion in `FormatterContentBuilder.prototype.appendToken()`.

This is benign in the context of formatting a standalone script file because `JSFormatter`
uses the original line endings only from the script text. But when formatting a HTML file with an inline
script, the formatter switches modes (`HTMLFormatter.prototype._formatWithNestedFormatter()`)
and the `JSFormatter` instance uses the original line endings from the HTML text.
This causes an offset for formatted positions of anything following
the multi-line template literal where its newlines aren&apos;t accounted for.

This patch updates handling of multi-line template literals the same way as a block comments,
inserting their contents line-by-line to map formatted to original line endings accordingly.

Tests: inspector/formatting/source-map-html-3.html
       inspector/formatting/source-map-html-4.html
       inspector/formatting/source-map-javascript-3.html
* LayoutTests/inspector/formatting/resources/html-source-map-tests/3.html: Added.
* LayoutTests/inspector/formatting/resources/html-source-map-tests/4.html: Added.
* LayoutTests/inspector/formatting/resources/html-tests/inline-script-expected.html:
* LayoutTests/inspector/formatting/resources/html-tests/inline-script.html:
* LayoutTests/inspector/formatting/resources/javascript-source-map-tests/3.js: Added.
* LayoutTests/inspector/formatting/resources/javascript-tests/template-strings-expected.js:
* LayoutTests/inspector/formatting/resources/javascript-tests/template-strings.js:
* LayoutTests/inspector/formatting/source-map-html-3-expected.txt: Added.
* LayoutTests/inspector/formatting/source-map-html-3.html: Added.
* LayoutTests/inspector/formatting/source-map-html-4-expected.txt: Added.
* LayoutTests/inspector/formatting/source-map-html-4.html: Added.
* LayoutTests/inspector/formatting/source-map-javascript-3-expected.txt: Added.
* LayoutTests/inspector/formatting/source-map-javascript-3.html: Added.
* Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
(JSFormatter.prototype._insertCommentsAndNewlines):
(JSFormatter.prototype._advanceLineEndingsIndexToEndOfToken):
Drive-by fix for a bug where multiple newlines are added before the token following a multi-line template literal,
one for each unaccounted newline in the template literal.

(JSFormatter.prototype._handleTokenAtNode):

Canonical link: <a href="https://commits.webkit.org/303680@main">https://commits.webkit.org/303680@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e746de30d67387616ba879065f14079825affc24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133327 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44461 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140882 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5693 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101989 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136274 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/119512 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82781 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1970 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113470 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37624 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143529 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5498 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38206 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110362 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5580 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4722 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110543 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28006 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4243 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115767 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5553 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5399 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69005 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5642 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
```diff
@@ -857,6 +866,12 @@ JSFormatter = class JSFormatter
             return;
         }
 
+        if ((nodeType === "TemplateLiteral" || nodeType === "TemplateElement") && tokenValue && tokenValue.includes("\n")) {
```
`&& tokenValue?.includes("\n")`

dcrousso replied:
i think to match what we do elsewhere, we should remove the checks for `"TemplateLiteral"` and `"TemplateElement"` below so that we dont have the same thing more than once
```js
if (nodeType === "TemplateLiteral" || nodeType === "TemplateElement") {
    if (tokenValue?.includes("\n")) {
        builder.appendStringWithPossibleNewlines(tokenValue, tokenOffset);
        this._advanceLineEndingsIndexToEndOfToken(token);
    } else
        builder.appendToken(tokenValue, tokenOffset);
    return;
}

// Include these here so we get only get warnings about unhandled nodes.
if (nodeType === "ExpressionStatement"
    || nodeType === "SpreadElement"
    || nodeType === "ImportAttribute"
    || nodeType === "ImportExpression"
    || nodeType === "Super"
    || nodeType === "MetaProperty"
    || nodeType === "RestElement"
    || nodeType === "DebuggerStatement"
    || nodeType === "AssignmentPattern"
    || nodeType === "ChainExpression") {
    builder.appendToken(tokenValue, tokenOffset);
    return;
}
```

dcrousso reviewed (approved):
are we sure that the results in `LayoutTests/inspector/formatting/source-map-html-3-expected.txt` are correct?  from a quick skim it seems like it, but evaluating +9,000 lines is a lot 😅

rcaliman-apple committed:
Web Inspector: Wrong breakpoint and search result positions after pretty-printing inline script with multi-line template literal
https://bugs.webkit.org/show_bug.cgi?id=291140
rdar://29417859

Reviewed by Devin Rousso.

`JSFormatter.prototype._handleTokenAtNode()` appended multi-line template literals
as a single token to the formatted content. This assumes the token is on a single line.

This actually triggered the assertion in `FormatterContentBuilder.prototype.appendToken()`.

This is benign in the context of formatting a standalone script file because `JSFormatter`
uses the original line endings only from the script text. But when formatting a HTML file with an inline
script, the formatter switches modes (`HTMLFormatter.prototype._formatWithNestedFormatter()`)
and the `JSFormatter` instance uses the original line endings from the HTML text.
This causes an offset for formatted positions of anything following
the multi-line template literal where its newlines aren't accounted for.

This patch updates handling of multi-line template literals the same way as a block comments,
inserting their contents line-by-line to map formatted to original line endings accordingly.

Tests: inspector/formatting/source-map-html-3.html
       inspector/formatting/source-map-html-4.html
       inspector/formatting/source-map-javascript-3.html
* LayoutTests/inspector/formatting/resources/html-source-map-tests/3.html: Added.
* LayoutTests/inspector/formatting/resources/html-source-map-tests/4.html: Added.
* LayoutTests/inspector/formatting/resources/html-tests/inline-script-expected.html:
* LayoutTests/inspector/formatting/resources/html-tests/inline-script.html:
* LayoutTests/inspector/formatting/resources/javascript-source-map-tests/3.js: Added.
* LayoutTests/inspector/formatting/resources/javascript-tests/template-strings-expected.js:
* LayoutTests/inspector/formatting/resources/javascript-tests/template-strings.js:
* LayoutTests/inspector/formatting/source-map-html-3-expected.txt: Added.
* LayoutTests/inspector/formatting/source-map-html-3.html: Added.
* LayoutTests/inspector/formatting/source-map-html-4-expected.txt: Added.
* LayoutTests/inspector/formatting/source-map-html-4.html: Added.
* LayoutTests/inspector/formatting/source-map-javascript-3-expected.txt: Added.
* LayoutTests/inspector/formatting/source-map-javascript-3.html: Added.
* Source/WebInspectorUI/UserInterface/Workers/Formatter/JSFormatter.js:
(JSFormatter.prototype._insertCommentsAndNewlines):
(JSFormatter.prototype._advanceLineEndingsIndexToEndOfToken):
Drive-by fix for a bug where multiple newlines are added before the token following a multi-line template literal,
one for each unaccounted newline in the template literal.

(JSFormatter.prototype._handleTokenAtNode):

Canonical link: https://commits.webkit.org/303680@main
